// eslint-plugin-testing-library/rules/await-fire-event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
const utils_1 = require("@typescript-eslint/utils");
const create_testing_library_rule_1 = require("../create-testing-library-rule");
const node_utils_1 = require("../node-utils");
exports.RULE_NAME = 'await-fire-event';
exports.default = (0, create_testing_library_rule_1.createTestingLibraryRule)({
    name: exports.RULE_NAME,
    meta: {
        type: 'problem',
        docs: {
            description: 'Enforce promises from `fireEvent` methods to be handled',
            recommendedConfig: {
                dom: false,
                angular: false,
                react: false,
                vue: 'error',
                marko: 'error',
            },
        },
        messages: {
            awaitFireEvent: 'Promise returned from `fireEvent.{{ name }}` must be handled',
            fireEventWrapper: 'Promise returned from `{{ name }}` wrapper over fire event method must be handled',
        },
